source/git/gitsource_unix.go

```diff
@@ -7,9 +7,16 @@ import (
 	"os/exec"
 	"syscall"
 	"time"
+
+	"golang.org/x/sys/unix"
 )
 
 func runProcessGroup(ctx context.Context, cmd *exec.Cmd) error {
+	{
+		// Need standard user umask for this process
+		old := unix.Umask(0022)
```

Comment:
This can leak the wrong umask for the other active goroutines and mess up the extraction if it happens at the same time. We should find some solution that doesn't switch main process umask. `--shared`, reexec or maybe something better.

Reply:
Yes it will leak. I tried with reexec but no luck, the buildkitd daemon is locked when I register:

```
#1 [internal] load git source https://github.com/crazy-max/buildkit-git-umask.git
#1 0.089 Initialized empty Git repository in /var/lib/buildkit/runc-overlayfs/snapshots/snapshots/1/fs/
#1 0.131 buildkitd: could not lock /var/lib/buildkit/buildkitd.lock, another instance running?
#1 0.131 main.main.func3
#1 0.131        /src/cmd/buildkitd/main.go:259
#1 0.131 github.com/urfave/cli.HandleAction
#1 0.131        /src/vendor/github.com/urfave/cli/app.go:523
#1 0.131 github.com/urfave/cli.(*App).Run
#1 0.131        /src/vendor/github.com/urfave/cli/app.go:285
#1 0.131 main.main
#1 0.131        /src/cmd/buildkitd/main.go:316
#1 0.131 runtime.main
#1 0.131        /usr/local/go/src/runtime/proc.go:255
#1 0.131 runtime.goexit
#1 0.131        /usr/local/go/src/runtime/asm_amd64.s:1581
#1 ERROR: failed to init repo at /var/lib/buildkit/runc-overlayfs/snapshots/snapshots/1/fs: exit status 1
```

I could probably put a condition on the lock but it seems disproportionate to our need. So I rewrote the git command by encapsulating it in a new shell. I don't really like it but I don't see any other solution that requires a minimum of changes.